### PR TITLE
codec: implement codec v2

### PR DESCRIPTION
**Pull request by jzelinskie**
This is a follow up from https://github.com/grpc/grpc-go/issues/6619

Go's gRPC implementation added a new `CodecV2` interface that enables integration with their memory pooling logic.
Adopting this should vastly improve performance and garbage collection overhead.

I took a stab at an implementation using the upstream version as reference.

In a review I'm looking for confirmation that this is the right approach for pooling and any feedback to make this more robust for a variety of vtprotobuf users. For example, I'm requiring `SizeVT()` here, which not everyone might generate; perhaps I could do a type assertion for that usage.

### CODE REVIEW

**Comment by mattrobenolt** [comment]
The downside I see to this implementation and the CodecV2 implementation in general is now the hard reliance on `grpc` the package since the interface relies on `mem.BufferSlice`.

This is kinda not ideal when using protobuf in contexts other than gRPC and use codecs for marshaling and unmarshalling. For example we also use vtprotobuf within ConnectRPC (https://connect.build) which effectively is an alternative to gRPC, so is kinda wonky to bring in a dependency on gRPC itself just to utilize this `mem.BufferSlice`.

It seems in general more appropriate to me if this `BufferSlice` were an interface, or a part of the core protobuf package rather than grpc itself since codecs aren't specific to gRPC and are often used outside of it even outside the context of RPC at all.

Prior to this, vtprotobuf has no runtime dependency on the grpc package, it's only used within the code generator, so this change makes that an explicit runtime depenedency which is awkward if you dont' even use grpc.

**Comment by jzelinskie** [comment]
I've typically vendored this Codec override into my codebase, but I'm PRing here for review. Maybe this could just be in an examples directory and not directly in vtprotobuf to avoid the dependency?

**Comment by coxley** [comment]
@jzelinskie: I don't personally have a strong opinion, but the example approach would fit given that's what is done for supporting both vtproto and standard in a single codec.

**Comment by coxley** [comment]
Also, while someone can generate only `SizeVT()`, they cannot use `MarshalVT()` without it. The generated marshaling code assumes size exists.

File: codec/grpc/grpc_codec.go [hunk 1]
```diff
@@ -1,33 +1,65 @@
...
+
+func (c *Codec) Marshal(v any) (data mem.BufferSlice, err error) {
+	if m, ok := v.(vtprotoMessage); ok {
+		size := m.SizeVT()
+		if mem.IsBelowBufferPoolingThreshold(size) {
+			buf := make([]byte, 0, size)
+			if _, err := m.MarshalToVT(buf[:0]); err != nil {
+				return nil, err
+			}
+			data = append(data, mem.SliceBuffer(buf))
```
Comment (na--) [inline codec/grpc/grpc_codec.go:35]: I was testing this patch and it returned empty (len=0, cap=size) buffers that failed some of my app tests. In my case the the underlying generated code for `MarshalToVT()` looks like this:
```go
func (m *Something) MarshalToVT(dAtA []byte) (int, error) {
	size := m.SizeVT()
	return m.MarshalToSizedBufferVT(dAtA[:size])
}
```

So I switched the desired `vtprotoMessage` interface to require `MarshalToSizedBufferVT` instead of `MarshalToVT`. We already have the result of `m.SizeVT()` anyway, which is not a simple thing to calculate, so this both fixes the bug and results in a more optimal code:
```suggestion
			buf := make([]byte, size)
			if _, err := m.MarshalToSizedBufferVT(buf); err != nil {
				return nil, err
			}
			data = mem.BufferSlice{mem.SliceBuffer(buf)}
```

Finally, there is no reason to use `append()` for the result. It will probably be optimized away by the Go compiler considering data is always going to be an empty slice here and below, but it's unnecessarily confusing when you read the code.

File: codec/grpc/grpc_codec.go [hunk 2]
```diff
@@ -1,33 +1,65 @@
...
+		} else {
+			pool := mem.DefaultBufferPool()
+			buf := pool.Get(size)
+			if _, err := m.MarshalToVT((*buf)[:0]); err != nil {
+				pool.Put(buf)
+				return nil, err
+			}
+			data = append(data, mem.NewBuffer(buf, pool))
```
Comment (na--) [inline codec/grpc/grpc_codec.go:43]: I haven't tested this yet, but similarly to the case above, I expect something like this would work better:
```suggestion
			if _, err := m.MarshalToSizedBufferVT((*buf)[:size]); err != nil {
				pool.Put(buf)
				return nil, err
			}
			data = mem.BufferSlice{mem.NewBuffer(buf, pool)}
```

**Comment by vmg** [comment]
Cool! First off, thanks @jzelinskie for taking the time to contribute and review this with us. @mattrobenolt's concern is certainly valid, but as you've already pointed out, the pattern we use for these kind of adapters is merging them as examples as to not bring specific dependencies (GRPC in this case) to the project. We've done something similar for DRPC and found it generally useful without requiring to bring DRPC as a dependency.

Now, for those following at home, the new buffers feature that shipped in GRPC 1.66 actually has a performance regression that fortunately @coxley has figured out and fixed (https://github.com/grpc/grpc-go/pull/7571). So thanks for that Codey! The fix hasn't been merged yet, which means that we can discuss the implementation of this side of the codec but running any actual benchmarks will be pointless until the fix lands and a new minor release is tagged.

As for the implementation itself: I think @na--'s comments are on point. `MarshalToSizedBufferVT` seems like the right API to use here, particularly since it handles the case of returning empty buffers in the same way as the previous codec and prevents the double-calculation of `SizeVT` (important!).

Besides those changes, we need to look at benchmarks before we can discuss further. I honestly have no idea of whether this will be a performance improvement at all! I think the killer feature here would be having a separate codegen step that allows serializing directly into `mem.BufferSlice`, because that could take very large messages (e.g. messages that have very large `[]byte` or `string` fields) and include those oversized parts of the message as individual buffers in the buffer slice, instead of having to allocate a massive single buffer and copying  the large `[]bytes` into it. Of course, that comes with its own nightmare of complexities and lifetimes which makes it a non-trivial problem, and I'm not sure whether that's something that can be generalized with a codegen step.

In conclusion, my plan here is as follows:

1. Wait for GRPC 1.66.1
2. Try using this codec in Vitess itself and measure performance impact.
3. If the codec is faster, try a manual implementation of "large []byte fields from ProtoBuf get passed directly as part of the `BufferSlice` for Vitess' MySQL packet ProtoBufs (which are essentially just a bunch of large `[]byte`).
4. If manually doing this kind of serialization has a measurable effect in benchmarks, try making a codegen phase for `vtprotobuf` that does just that for any oversized byte fields it finds during serialization.

I'll keep y'all posted!

**Comment by coxley** [comment]
@vmg Something that may be neat is finally being able to use pooled vtproto objects for unary client and servers. 

It’d be a bit wonky — it would rely on a custom BufferPool. If the pool and codec could communicate, the uintptr for a pooled slice could be mapped to a vtproto so that on pool.Put, we know we can release it. 

May be more indirection than is worth it, but the pool interface at least makes it possible to derive the lifetime from gRPCs perspective.

**Comment by vmg** [comment]
Now testing https://github.com/vitessio/vitess/pull/16790 in Vitess itself. Hoping this will show up in the benchmarks!

File: codec/grpc/grpc_codec.go [hunk 3]
```diff
@@ -1,33 +1,63 @@
...
-	vt, ok := v.(vtprotoMessage)
-	if !ok {
-		return nil, fmt.Errorf("failed to marshal, message is %T (missing vtprotobuf helpers)", v)
+func (Codec) Name() string { return Name }
+
+func (c *Codec) Marshal(v any) (mem.BufferSlice, error) {
+	if m, ok := v.(vtprotoMessage); ok {
+		size := m.SizeVT()
+		if mem.IsBelowBufferPoolingThreshold(size) {
+			buf := make([]byte, 0, size)
```
Comment (coufalja) [inline codec/grpc/grpc_codec.go:31]: There is a bug on this line IMO and it should go like either `make([]byte, size)` or  `mem.BufferSlice{mem.SliceBuffer(buf[:size])}` on the line 35. Because the original slice reference will have its len still set to 0 effectively returning 0 length byte array from the method.

Comment (jzelinskie) [reply codec/grpc/grpc_codec.go:31]: Yes! Apologies -- I have this fixed in my [PR for SpiceDB](https://github.com/authzed/spicedb/pull/2070) but didn't carry over the changes here.

**Comment by vmg** [comment]
This has worked wonderfully in upstream Vitess: https://github.com/vitessio/vitess/pull/16790#issuecomment-2419021276

**Comment by francoposa** [comment]
@vmg I just came across this after attempting to hack bufferpooling into csproto - is there a timeline or likelihood of getting this merged / released ?

Thanks!

File: codec/grpc/grpc_codec.go [hunk 4]
```diff
@@ -1,33 +1,63 @@
...
-	return vt.UnmarshalVT(data)
+
+	return c.fallback.Unmarshal(data, v)
 }
 
-func (Codec) Name() string {
-	return Name
+func init() {
+	encoding.RegisterCodecV2(&Codec{
+		fallback: encoding.GetCodecV2("proto"),
```
Comment (janisz) [inline codec/grpc/grpc_codec.go:61]: We can use name from proto so we will avoid unused import https://pkg.go.dev/google.golang.org/grpc/encoding/proto#pkg-constants
```suggestion
		fallback: encoding.GetCodecV2(proto.Name),
```

**Comment by vmg** [comment]
@francoposa: What are you exactly looking for in a release? As explained earlier, we're not looking to merge the custom GRPC codec because that'd introduce a GRPC dependency on the project, so the plan for interested users is to implement the coded themselves. We've done so in Vitess and it was very successful. Besides the pooling, which has worked nicely, the custom serializer into chunks has been proving to be very complex to implement with very few performance benefits to show for it.
